src/bundle/Security/Http/EventListener/CheckTwoFactorCodeReuseListener.php

```diff
@@ -0,0 +1,65 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Scheb\TwoFactorBundle\Security\Http\EventListener;
+
+use Psr\Cache\CacheItemInterface;
+use Psr\Cache\CacheItemPoolInterface;
+use Psr\EventDispatcher\EventDispatcherInterface;
+use Scheb\TwoFactorBundle\Security\TwoFactor\Event\TwoFactorCodeCheckEvent;
+use Scheb\TwoFactorBundle\Security\TwoFactor\Event\TwoFactorCodeReusedEvent;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use function assert;
+use function sha1;
+
+/**
+ * @final
+ */
+class CheckTwoFactorCodeReuseListener implements EventSubscriberInterface
+{
+    public const LISTENER_PRIORITY = 0;
+
+    public function __construct(
+        private readonly EventDispatcherInterface $eventDispatcher,
+        private readonly mixed $cache = null,
+    ) {
+    }
+
+    public function checkForCodeReuse(TwoFactorCodeCheckEvent $event): void
+    {
+        if (null === $this->cache) {
+            return;
+        }
+
+        $cacheItem = null;
+        assert($cacheItem instanceof CacheItemInterface || null === $cacheItem);
+        $cacheKey = 'scheb_two_factor_code_reuse.'
+            .sha1($event->getUser()->getUserIdentifier().'.'.$event->getCode());
+
+        if ($this->cache instanceof CacheItemPoolInterface) {
+            $cacheItem = $this->cache->getItem($cacheKey);
+            $cacheItem->expiresAfter(60);
```

Comment:
The 60 seconds are an arbitrary number. But as most tokens are configured to invalidate after 30 seconds that should be enough. If need be we could also make this configurable but the less moving parts the less confusing it is for the users.

Reply:
Implemented